cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/SymbolResolver.kt

```diff
@@ -254,6 +256,41 @@ open class SymbolResolver(ctx: TranslationContext) : ComponentPass(ctx) {
         }
     }
 
+    /**
+     * This function tries to resolve a [Node.name] to a list of symbols (a symbol represented by a
+     * [Declaration]) starting with [startScope]. This function can return a list of multiple
+     * symbols in order to check for things like function overloading. but it will only return list
+     * of symbols within the same scope; the list cannot be spread across different scopes.
+     *
+     * This means that as soon one or more symbols are found in a "local" scope, these shadow all
+     * other occurrences of the same / symbol in a "higher" scope and only the ones from the lower
+     * ones will be returned.
+     */
+    private fun findSymbols(
+        nodeWithName: Node,
+        startScope: Scope? = scopeManager.currentScope
+    ): Set<Declaration> {
+        val (scope, name) = scopeManager.extractScope(nodeWithName, startScope)
+        val list =
+            scopeManager
+                .resolve<Declaration>(scope, true) { it.name.lastPartsMatch(name) }
+                .toMutableSet()
+        // If we have both the definition and the declaration of a function declaration in our list,
+        // we chose only the definition
+        val it = list.iterator()
```

Comment:
Is this faster than using a filter like `list.filter {...}` ?

Reply:
Hmm it is probably a little bit faster since we are removing elements from the list that we need to check, so if we have a large list of function declarations, using the mutable iterator should be faster since we are removing those declarations step by step. Using the filter on the original list would mean that we need to loop through the original list items even though we know that the element was already removed.

Reply:
Ah there is the check if its definition is already in the list so we cannot merge the entire conditional removing into the filter or predicate of the scope resolution above. At least not trivially